server/src/main/java/io/druid/segment/realtime/firehose/EventReceiverFirehoseFactory.java

```diff
@@ -264,11 +257,29 @@ public int getCapacity()
     @Override
     public void close() throws IOException
     {
-      log.info("Firehose closing.");
-      closed = true;
-      eventReceiverFirehoseRegister.unregister(serviceName);
-      if (chatHandlerProvider.isPresent()) {
-        chatHandlerProvider.get().unregister(serviceName);
+      if (!closed) {
+        log.info("Firehose closing.");
+        closed = true;
```

Comment:
Can you explain a bit more about how this makes certain we don't have events in flight which get dropped in the process of closing?